lib/Doctrine/DBAL/TransactionManager.php

```diff
@@ -0,0 +1,275 @@
+
+    /**
+     * Whether nested transactions should use savepoints.
+     *
+     * @var boolean
+     */
+    private $nestTransactionsWithSavepoints = false;
+
+    /**
+     * @param \Doctrine\DBAL\Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @return TransactionBuilder
+     */
+    public function createTransaction()
+    {
+        return new TransactionBuilder($this);
+    }
+
+    /**
+     * @param array $configuration
+     *
+     * @return \Doctrine\DBAL\Transaction
+     */
+    public function beginTransaction(array $configuration = array())
+    {
+        $this->connection->connect();
+
+        if (isset($configuration[TransactionBuilder::ISOLATION_LEVEL])) {
+            $this->connection->setTransactionIsolation($configuration[TransactionBuilder::ISOLATION_LEVEL]);
+        }
+
+        $transaction = new Transaction($this, $configuration);
+        $this->activeTransactions[] = $transaction;
+
+        $logger = $this->connection->getConfiguration()->getSQLLogger();
+
+        if (count($this->activeTransactions) === 1) {
+            $logger && $logger->startQuery('"START TRANSACTION"');
+            $this->connection->getWrappedConnection()->beginTransaction();
+            $logger && $logger->stopQuery();
+        } elseif ($this->nestTransactionsWithSavepoints) {
+            $logger && $logger->startQuery('"SAVEPOINT"');
+            $this->connection->createSavepoint($this->getNestedTransactionSavePointName());
+            $logger && $logger->stopQuery();
+        }
+
+        return $transaction;
+    }
+
+    /**
+     * @param \Doctrine\DBAL\Transaction $transaction
+     *
+     * @return void
+     *
+     * @throws \Doctrine\DBAL\ConnectionException If the transaction is stale, or belongs to another manager.
+     */
+    public function commitTransaction(Transaction $transaction)
+    {
+        if ($transaction->isActive()) {
+            // Transaction::commit() has to be called first, and in turn call this method.
+            $transaction->commit();
+
+            return;
+        }
+
+        $nestedTransaction = $this->getNestedTransaction($transaction);
+
+        if ($nestedTransaction) {
+            $nestedTransaction->commit();
+        }
+
+        $logger = $this->connection->getConfiguration()->getSQLLogger();
+
+        if (count($this->activeTransactions) === 1) {
+            $logger && $logger->startQuery('"COMMIT"');
+            $this->connection->getWrappedConnection()->commit();
+            $logger && $logger->stopQuery();
+        } elseif ($this->nestTransactionsWithSavepoints) {
+            $logger && $logger->startQuery('"RELEASE SAVEPOINT"');
+            $this->connection->releaseSavepoint($this->getNestedTransactionSavePointName());
+            $logger && $logger->stopQuery();
+        }
+
+        array_pop($this->activeTransactions);
```

Comment:
Note sure about this one though: if committing / rolling back fails, don't we want to leave the TXM unchanged, just like you suggest to do when *beginning* the transaction fails?